README
#In The Clear
In The Clear is an iOS directions app that shows the user the weather they can expect as they are driving. It not only shows the weather via the Google Maps polyline (directions path displayed on map), but also in a table that shows the weather in each city the user will be driving through. If you are ever driving over a long distance, especially if you're traveling through an area that commonly receives unfavorable weather, this app can be a great tool that can help you predict what weather conditions you can expect to drive through. 

#Requirements
- iOS 12.0+
- XCode 10.0+
- Swift 4.0

#Getting Started
##CocoaPods
- Add to your 'Podfile':

```ruby
platform :ios, '12.0'

target 'InTheClear' do
  use_frameworks!

  pod 'GoogleMaps', '~> 3.0.2'
  pod 'GooglePlaces', '~> 3.0.2'
  pod 'Alamofire', '~> 4.7'
  pod 'SwiftyJSON', '~> 4.0'
  pod 'FontAwesome.swift', '~> 1.6.0'
end
```

Then run 'pod install' you will have all the necessary CocoaPods installed.

##Setting up Google Maps, Google Places, and OpenWeatherAPI
To run this project you will need an API key for GoogleMaps, Google Places, and OpenWeatherAPI. Insert the API keys in the following locations

- Google Maps
In AppDelegate.swift:
```swift
var googleAPIKey = "Your-API-Key"
```

- Google Places
In AppDelegate.swift:
```swift
var googlePlacesKey = "Your-API-Key"
```

- OpenWeatherAPI
```swift
let urlComplete = urlBase + "lat=\(lat)&lon=\(long)&units=imperial&APPID=Your-API-Key"
```

All that's left to do is to open 'InTheClear.xcworkspace' and you should be good to go!


Changelog.md
# Changelog

[Version 1.1] - Released 7/14/19
* New backend API for reverse geolocation - 7/1/19
* New backend API for weather - 7/5/19
* Added weather images - 7/6/19
* Error handling for failed API call - 7/7/19
* Added severe weather icon and  route icon - 7/8/19
* Made Temperature and Condition labels the same size - 7/9/19
* Moved all API calls to a single file - 7/9/19
* Updated Location request string - 7/12/19

[Version 1.2] - Released 7/23/19
* Created time button for Set Time View - 7/17/19
* Added Set Time View - 7/18/19
* Added Night Time icons and cell colors to Weather Table View - 7/19/19
* Showed Set Time View when direction input is given - 7/19/19
* Added Location button customized location button - 7/20/19
* Removed Location button if location services are not enabled - 7/21/19
* Increased time limit to 60 hours - 7/21/19

[Version 2.0] - Released 8/20/19
* File Structure moved over to MVP structure - 7/26/2019
* Ability to add city stops - 8/3/19
* Alowed reorder of city stops - 8/3/19
* Map shows each stop - 8/5/19
* Allowed for preset departure times - 8/10/19
* Redesign trip time/distance label - 8/14/19
* Allowed waypoints to be added to Google Maps URL - 8/14/19
* New logo added to splash screen - 8/14/19
* Swift 5 update - 8/16/19
* Access types added to functions - 8/17/19
* Google Places table darkened and iOS 13 support added but commented out - 8/19/19

[Version 2.1] - Released 10/1/19
* Added supported for iOS 13 dark mode - 9/23/19

[Version 3.0] - Released 3/17/20
* Added Storyboard for Sign in - 1/12/20
* Added Apple Sign in - 1/12/20
* Added Google Sign in - 2/1/20
* Login Data Route Added - 2/6/20
* Added Posting of Trip Data - 2/8/20
* Added Firebase Authentication - 2/11/20
* Added Realm for User Data and Token Storing - 2/14/20
* Login Route Updated for Firebase - 2/14/20
* Increased Trip Leg Time Limit to 48 hours - 3/1/20

[Version 3.1] - Released 3/12/22
* Added Codefactor for code analysis - 8/13/20
* Added Storekit Handler - 12/1/21
* Added Premium Membership - 12/1/21
* Profile Action Table Added - 12/7/21
* Free tier limits Added - 12/12/21
* Added Firestore - 12/18/21
* Added Premium Ad - 12/18/21
* Added Restore for Purchases - 2/8/22

[Version 3.2]
* UI/UX Enhancements - 3/19/22
* Added option to open route in Apple Maps - 3/22/22
* Route selector bug fix - 3/26/22


Weather.swift
//
//  Weather.swift
//  InTheClear
//
//  Created by Josh Sauder on 9/19/18.
//  Copyright © 2018 Josh Sauder. All rights reserved.
//

import Foundation
import Alamofire
import SwiftyJSON
import GoogleMaps
import MapKit

extension ViewController {
    
    struct pathColorSegs {
        static let SNOW = GMSStrokeStyle.solidColor(UIColor(red:0.12, green:0.53, blue:0.90, alpha:1.0))
        static let RAIN = GMSStrokeStyle.solidColor(UIColor(red:0.26, green:0.63, blue:0.28, alpha:1.0))
        static let STORMS = GMSStrokeStyle.solidColor(UIColor(red:0.90, green:0.22, blue:0.21, alpha:1.0))
        static let SUN = GMSStrokeStyle.solidColor(UIColor(red:0.98, green:0.75, blue:0.18, alpha:1.0))
        static let CLOUDS = GMSStrokeStyle.solidColor(UIColor(red:0.38, green:0.49, blue:0.55, alpha:1.0))
    }
    
    /**
     Creates an alert on error on the users device
     Stops the spinner and moves the camera back to the users start location
     
     - parameters:
        - title: The title of the alert
        - message: The message contained in the alert
 `  */
    func alertTool(title: String, message: String){
        self.showAlert(title: title, message: message)
        
        //animate back to start location
        let camera = GMSCameraPosition.camera(withLatitude: (self.locationStart.coordinate.latitude), longitude: (self.locationStart.coordinate.longitude), zoom: 11.0)
        self.mapView.animate(to: camera)
        
        //need to stop spinner since completion will not be used
        self.stopSpinner()
        //re-enable location buttons
        self.startButton.isEnabled = true
        self.destinationButton.isEnabled = true
    }
    
    /**
     Creates the parameters needed for both API requests
     
     - parameters:
        - steps: The directions JSON array
     - returns: Two sets of parameters for the two AWS Lambda functions
    */
    func createParamters(steps: [JSON], date: Date) -> (Parameters, Parameters){
        
        //get the time for the current weather step
        let refDate = date.timeIntervalSince1970
        let timeInterval = TimeInterval(refDate)
        var date = Date(timeIntervalSince1970: timeInterval)
        
        var weatherArray: [[String: Any]] = []
        var geolocationArray: [[String: Any]] = []
        
        for step in steps {
            var dictionaryItem: [String: Any] = [:]
            
            //add latitude and longitude coordinates to dictionary
            dictionaryItem["lat"] = step["end_location"]["lat"].stringValue
            dictionaryItem["long"] = step["end_location"]["lng"].stringValue
        
            geolocationArray.append(dictionaryItem)
            
            //time in seconds
            let stepTime = step["duration"]["value"].intValue
            date = date.addingTimeInterval(TimeInterval(stepTime))
            dictionaryItem["time"] = date.timeIntervalSince1970.rounded()
            
            weatherArray.append(dictionaryItem)
        }
        
        let weatherParameters: Parameters = ["List" : weatherArray]
        let geolocationParamters: Parameters = ["List": geolocationArray]
        return (weatherParameters, geolocationParamters)
        
    }
    
    
    /**
     Creates GMSPath on the Map
     
     - parameters:
        - startLocation: The starting location
        - endLocation: The destination location
        - completion: Upon calling the service, return the total time and distance strings
    */
    func createLine(startLocation: CLLocation, endLocation: CLLocation, date: Date, completion: @escaping (String, String) -> ()) {
        
        //colors for lines based on condition
        
        let origin = "\(startLocation.coordinate.latitude),\(startLocation.coordinate.longitude)"
        let destination = "\(endLocation.coordinate.latitude),\(endLocation.coordinate.longitude)"
        
        let pathURL = url.PATH_URL + origin + "&destination=" + destination + "&mode=driving&key=" + Constants.GOOGLE_DIRECTIONS_KEY
        
        self.getDirections(url: pathURL) { routes in
            
            //prevents invalid routes from being inputed
            if routes.count > 0 {
                
                //get the direction steps from the directions array
                let routesVal = routes[0]["legs"].arrayValue
                let stepsEval = routesVal[0]
                let totalTime = stepsEval["duration"]["text"].stringValue
                let totalDistance = stepsEval["distance"]["text"].stringValue
                let steps = stepsEval["steps"].arrayValue
                
                //prevent routes that are too long too handle
                if stepsEval["duration"]["value"].intValue < 216000 {
                //remove any existing polylines
                    self.polylineArray.forEach { $0.map = nil }

                    
                    //take first route and use polyline to draw line
                    let route = routes[0]
                    let routeOverviewPolyline = route["overview_polyline"].dictionary
                    let points = routeOverviewPolyline?["points"]?.stringValue
                    
                    //create path and polyline from encoded path
                    let path = GMSPath.init(fromEncodedPath: points!)
                    let polyline = GMSPolyline.init(path: path)
                    polyline.strokeWidth = 7
                    
                    //make sure geolocating is complete when Polyline is colored
                    let group = DispatchGroup()
                    group.enter()
                    group.enter()
                    
                    let (weatherParam, geolocationParam) = self.createParamters(steps: steps, date: date)
                    //let geolocating and path coloring run simultaneouly
                    self.getLocationName(parameters: geolocationParam){
                        group.leave()
                    }
                    //color the polyline and on completion show polyline on map
                    self.colorPath(line: polyline, steps: steps, path: path!, parameters: weatherParam) {
                        if(self.conditions.count > 0){
                            polyline.map = self.mapView
                            self.polylineArray.append(polyline)
                            self.mapView.animate(with: GMSCameraUpdate.fit(GMSCoordinateBounds(path: polyline.path!), withPadding: 95))
                        }
                        group.leave()
                    }
                    
                    group.notify(queue: DispatchQueue.main){
                        //return total time value from json once colorpath and geolocating methods are complete
                        completion(totalTime, totalDistance)
                    }
                }
                else{
                    let title = "Invalid Route"
                    let message = "Sorry! At this time, In The Clear does not support routes longer than 60 hours."
                    
                    self.alertTool(title: title, message: message)
                }
            }
            else {
                let title = "Invalid Route"
                let message = "Woops! Looks like it's not possible to drive between these two locations."
                
                self.alertTool(title: title, message: message)
            }
        }
    }

    
    /**
     Colors the GMSPath by what the weather condition is at each coordinate
     
     - parameters:
        - line: The GMSPolyLine representation of the path
        - steps: Each direction (step) from the JSON returned by the Google Directions API
        - path: The directions path
        - parameters: The API request body
        - completion: After weather services callback, exit function
    */
    func colorPath(line: GMSPolyline, steps: [JSON], path: GMSPath, parameters: Parameters, completion: @escaping () -> ()) {
        //take each step and get weather at end location
        var colorSegs: [GMSStyleSpan] = []
    
        weatherPerStep(steps: steps, path: path, parameters: parameters){ result in
            //on completion get the colorSegs array and set it equal to the style span of the polyline
            colorSegs =  result
            line.spans = colorSegs
            completion()
        }
        
    }
    
    /**
     Gets the weather at each directions step and colors the GMSPath accordingly
     - parameters:
        - steps: The array containing each directions step
        - path: The GMSPath that will be displayed on the map
        - parameters: The API request body
        - completion: An array containing the color segments for the polyline.
    */
    func weatherPerStep(steps: [JSON], path: GMSPath, parameters: Parameters, completion: @escaping ([GMSStyleSpan]) -> ()) {
        
        var colorSegs: [GMSStyleSpan] = []

        getWeather(parameters: parameters){ json in
            
            var i = UInt(0)
            for (index, item) in json.enumerated() {
                let condition = item["Condition"] as! String
                //determine number of segs for each step
                let (numSegs, index) = self.determineSegCount(step: steps[index], path: path, index: i)
                i = index
                //append each value to corresponding array
                colorSegs.append(self.determineColorSeg(condition: condition, numberSegs: numSegs))
                if item["Severe"] as! Bool {
                    self.conditions.append("danger")
                } else{
                    self.conditions.append(condition)
                }
                self.conditionDescription.append(item["Description"] as! String)
                let temp = item["Temperature"] as! NSNumber
                self.highTemps.append(temp.floatValue)
            }
            //return colorsegs on completion
            completion(colorSegs)
        }
    }
    
    /**
     Determines the number of segments on the GMSPolyline each step requires
     - parameters:
        - steps: The single directions step
        - path: The path displayed on the users device
        - index: the coordinate index on the GMSPath
     - returns: The number of segs and the last segment index on the GMSPath
    */
    func determineSegCount(step: JSON, path: GMSPath, index: UInt) -> (Int, UInt){
        //array will contain path index and number of segs
        var i = index
        //get path index
        var pathCoordinates = path.coordinate(at: i)
        //determine step location
        let stepCoordinates = CLLocationCoordinate2D(latitude: step["end_location"]["lat"].doubleValue, longitude: step["end_location"]["lng"].doubleValue)
        
        //while path is in range of step add segs
        var numberSegs = 1
        //start from start and go to end... since using end for path
        while abs(pathCoordinates.latitude - stepCoordinates.latitude) > 0.3 || abs(pathCoordinates.longitude - stepCoordinates.longitude) > 0.3 {
            //increment the number of segs and the path counter
            numberSegs += 1
            i += 1
            pathCoordinates = path.coordinate(at: i)
        }
        
        return (numberSegs, i)
    }
    
    
    /**
     Determines what color each segment in polyline should be
     
     - parameters:
        - condition: The weather condition for the segment(s)
        - numberSegs: The number of segments that need to be colored
     - returns: The GMSStyleSpan for a specific segment(s)
     
    */
    func determineColorSeg(condition: String, numberSegs: Int) -> GMSStyleSpan {
        let colorSeg: GMSStyleSpan
        //determine condition and set segment color accordingly
        if condition == "rain" {
            colorSeg = GMSStyleSpan(style: pathColorSegs.RAIN, segments: Double(numberSegs))
        } else if condition == "danger" {
            colorSeg = GMSStyleSpan(style: pathColorSegs.STORMS, segments: Double(numberSegs))
        } else if condition == "snow" || condition == "sleet" {
            colorSeg = GMSStyleSpan(style: pathColorSegs.SNOW, segments: Double(numberSegs))
        } else if condition == "cloudy" || condition == "partly-cloudy-day" || condition == "partly-cloudy-night" {
            colorSeg = GMSStyleSpan(style: pathColorSegs.CLOUDS, segments: Double(numberSegs))
        } else {
            colorSeg = GMSStyleSpan(style: pathColorSegs.SUN, segments: Double(numberSegs))
        }
        
        return colorSeg
    }
}
